ckanext/canada/assets/internal/canada_internal_analytics.js

```diff
@@ -0,0 +1,319 @@
+window.addEventListener('load', function(){
+  // Add Google Analytics Custom Event Tracking to actions on the Registry
+  $(document).ready(function() {
+    let user, title;
+    let referrer = document.referrer;
+
+    // Get the username from the Logged In toolbar
+    if( $('#usr-logged').length > 0 &&
+        $('#usr-logged').find('.username').length > 0
+    ){
+      user = $('#usr-logged').find('.username')[0].innerText;
+    }
+
+    // Get the title from the first H1 tag
+    if( $('h1').length > 0 ){
+      title = $('h1')[0].innerText;
+    }
+
+    if( typeof gtag != 'undefined' ){
+      // gtag function exists
+
+      let flashes = $('.flash-messages');
+      if( flashes.length > 0 &&
+          flashes[0].innerText.length > 0
+      ){
+        // there are flash messages with content in them
+
+        let messages = flashes[0].innerText; // get all text inside messages (excludes HTML tags)
+
+        // track successful login
+        if( messages.match(/.*now logged in.*/) ||
+            messages.match(/.*est maintenant connecté.*/)
+        ){
+
+          gtag('event', 'user', {
+            'action': 'Login Successful',
+            'user': user,
+            'title': title,
+            'referrer': referrer,
+            'messages': messages,
+          });
+
+        }
+
+        // track failed login
+        if( messages.match(/.*Login failed.*/) ||
+            messages.match(/.*Authentification échouée.*/)
+        ){
+
+          gtag('event', 'user', {
+            'action': 'Login Failed',
+            'user': user,
+            'title': title,
+            'referrer': referrer,
+            'messages': messages,
+          });
+
+        }
+
+        // track password reset request
+        if( messages.match(/.*check your e-mail inbox in order to confirm your password reset.*/) ||
+            messages.match(/.*consulter votre boîte à lettres électronique pour confirmer votre demande.*/)
```

Comment:
these messages sound like they're referring to different things

Reply:
also this pattern of matching text seems really fragile. Can we add some data attributes or something else that won't change over time?

Reply:
For the first comment: I just copied from our PO files for the French translations.

Second comment: Yeah, I also did not like how this was done very much. The front-end analytics stuff always gets tricky with back-end frameworks. But as the flash messages are session based, it makes sense to do it on the flash message I guess. As for the data attributes or classes, that would be wildly preferable. It is just that we would probably have to override a bunch of views to get what we want in the flash messages?

Unless if we can think of some super cool fancy way to catch flash messages and see what action called them? and then insert something there. E.g. the `package_create` action did `h.flash_success` and we could see which view method called `package_create` or something. And then just map CKAN actions to GA Custom Events

Reply:
we could override the `flash_success` and related helpers to capture information about the caller

Reply:
Yes, that would be the CORRECT way to do it. But what about throwing pythonian out the window and directly Monkey Patching the `_Flash` class from the core helpers :)

Reply:
That's an option if overriding the helpers seems like it would difficult to maintain, but likely better done in our ckan fork instead of monkey patching.